File: src/main/java/com/tuliomagalhaes/graph/adjacencymatrix/Graph.java
package com.tuliomagalhaes.graph.adjacencymatrix;

public class Graph<T> {
    private final int CONNECTED = 1;
    private final int DISCONNECTED = 0;

    private final int size;
    private final T[] vertexList;
    private final int[][] adjacencyMatrix;
    private int numberOfVertex;

    public Graph(int size) {
        this.size = size;
        numberOfVertex = 0;
        vertexList = (T[]) new Object[size];
        adjacencyMatrix = new int[size][size];

        initializeAdjacencyMatrix();
    }

    public void addVertex(T vertex) {
        vertexList[numberOfVertex++] = vertex;
    }

    public void addEdge(int vertex1, int vertex2) {
        adjacencyMatrix[vertex1][vertex2] = CONNECTED;
        adjacencyMatrix[vertex2][vertex1] = CONNECTED;
    }

    public T getVertexAt(int index) {
        return vertexList[index];
    }

    public boolean isConnected(int vertex1, int vertex2) {
        return adjacencyMatrix[vertex1][vertex2] == CONNECTED;
    }

    private void initializeAdjacencyMatrix() {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                adjacencyMatrix[i][j] = DISCONNECTED;
            }
        }
    }
}

File: src/main/java/com/tuliomagalhaes/graph/adjacencylist/Graph.kt
package com.tuliomagalhaes.graph.adjacencylist

import java.util.*

class Graph<T> {
    private val adjacencyList = mutableMapOf<T, MutableSet<T>>()

    fun addVertex(vertex: T) {
        if (!adjacencyList.containsKey(vertex)) {
            adjacencyList[vertex] = mutableSetOf()
        }
    }

    fun addEdge(vertex1: T, vertex2: T) {
        addVertex(vertex1)
        addVertex(vertex2)

        adjacencyList[vertex1]?.add(vertex2)
        adjacencyList[vertex2]?.add(vertex1)
    }

    fun containsVertex(vertex: T): Boolean {
        return adjacencyList[vertex] != null
    }

    fun isConnected(vertex1: T, vertex2: T): Boolean {
        return adjacencyList[vertex1]?.contains(vertex2) == true &&
               adjacencyList[vertex2]?.contains(vertex1) == true
    }

    fun depthFirstSearchRecursive(vertex: T): List<T> {
        val visitedVertices = mutableSetOf<T>()
        val result = mutableListOf<T>()

        fun dfs(vertex: T) {
            adjacencyList[vertex]?.let { neighbors ->
                visitedVertices.add(vertex)
                result.add(vertex)
                neighbors.forEach { neighbor ->
                    if (!visitedVertices.contains(neighbor)) {
                        dfs(neighbor)
                    }
                }
            }
        }
        dfs(vertex)

        return result
    }

    fun breadthFirstSearchRecursive(vertex: T): List<T> {
        val visitedVertices = mutableSetOf(vertex)
        val result = mutableListOf<T>()
        val nextVertices = LinkedList<T>()
        nextVertices.push(vertex)

        fun bfs(vertex: T) {
            if (nextVertices.isEmpty()) return
            val nextVertex = nextVertices.pop()
            result.add(nextVertex)
            adjacencyList[nextVertex]?.let { neighbors ->
                neighbors.forEach { neighbor ->
                    if (!visitedVertices.contains(neighbor)) {
                        visitedVertices.add(neighbor)
                        nextVertices.add(neighbor)
                    }
                }

                bfs(vertex)
            }
        }
        bfs(vertex)

        return result
    }

    fun depthFirstSearchIterative(vertex: T): List<T> {
        return traverse(vertex, true)
    }

    fun breadthFirstSearchIterative(vertex: T): List<T> {
        return traverse(vertex, false)
    }

    private fun traverse(vertex: T, dfs: Boolean): List<T> {
        if (!adjacencyList.containsKey(vertex)) return emptyList()

        val visitedVertices = mutableSetOf(vertex)
        val result = mutableListOf<T>()
        val nextVertices = LinkedList<T>()
        nextVertices.push(vertex)

        while (nextVertices.isNotEmpty()) {
            val nextVertex = nextVertices.pop()
            result.add(nextVertex)

            adjacencyList[nextVertex]?.forEach { neighbor ->
                if (!visitedVertices.contains(neighbor)) {
                    visitedVertices.add(neighbor)
                    if (dfs) {
                        nextVertices.push(neighbor)
                    } else {
                        nextVertices.add(neighbor)
                    }
                }
            }
        }

        return result
    }

    fun removeEdge(vertex1: T, vertex2: T) {
        adjacencyList[vertex1]?.remove(vertex2)
        adjacencyList[vertex2]?.remove(vertex1)
    }

    fun removeVertex(vertex: T) {
        adjacencyList[vertex]?.forEach {
            adjacencyList[it]?.remove(vertex)
        }
        adjacencyList.remove(vertex)
    }
}